### PR TITLE
Fix RTL3b and RTL4b.

### PR DESCRIPTION
**Pull request by tcard**
[Review on Reviewable](https://reviewable.io/reviews/ably/ably-ios/223)

### CODE REVIEW

File: ably-ios/ARTRealtime.m [hunk 1]
```diff
@@ -758,7 +768,11 @@ - (void)realtimeTransportClosed:(id<ARTRealtimeTransport>)transport {
 }
 
 - (void)realtimeTransportDisconnected:(id<ARTRealtimeTransport>)transport {
-    [self transition:ARTRealtimeDisconnected];
+    if (self.connection.state == ARTRealtimeClosing) {
+        [self transition:ARTRealtimeClosed];
```
Comment (ricardopereira) [inline ably-ios/ARTRealtime.m]: It needs to wait for the CLOSED action to transition to the Closed state. [RTN12a](http://docs.ably.io/client-lib-development-guide/features/#RTN12a)

Comment (tcard) [inline ably-ios/ARTRealtime.m]: But this is a disconnection. If there is a disconnection, and it was closing, the CLOSED action obviously can't arrive, but the client should act as if it happened. That's what I understand from [RTN12c](http://docs.ably.io/client-lib-development-guide/features/#RTN12c).

Comment (ricardopereira) [inline ably-ios/ARTRealtime.m]: Sorry, I missed the _ disconnection_, you right. :+1:

**Comment by ricardopereira** [comment]
LGTM

File: ably-ios/ARTRealtime.m [hunk 2]
```diff
@@ -434,6 +429,21 @@ - (void)onDisconnected {
...
+    [self.logger info:@"ARTRealtime closed"];
+    switch (self.connection.state) {
+        case ARTRealtimeClosed:
+            break;
+        case ARTRealtimeClosing:
+            [self.connection setId:nil];
+            [self transition:ARTRealtimeClosed];
+            break;
+        default:
+            NSAssert(false, @"Invalid Realtime state: expected Closing or Closed has current state");
```
Comment (mattheworiordan) [inline ably-ios/ARTRealtime.m]: Typo I think, not sure I understand the `has current state` part

**Comment by mattheworiordan** [comment]
One small comment, but otherwise :+1:

**Comment by tcard** [comment]
Review status: 0 of 4 files reviewed at latest revision, 1 unresolved discussion.

---

*[ably-ios/ARTRealtime.m, line 442 \[r1\]](https://reviewable.io:443/reviews/ably/ably-ios/223#-KApA29wn4kMQLEghDZJ-r1-442:-KApHPVTAvHEIEXCR5eC:672255914) ([raw file](https://github.com/ably/ably-ios/blob/9898b6340c90c1ba66c45839afe7f7538bf68bca/ably-ios/ARTRealtime.m#L442)):*
True... was there already, I just copypasted and didn't notice. Fixed at a60f848.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/ably/ably-ios/223)*

**Pull request merged**